prov/sm2/src/sm2_rma.c

```diff
@@ -0,0 +1,509 @@
+	ctx->msg.iov_count = msg->iov_count;
+	ctx->msg.addr = msg->addr;
+	ctx->msg.data = msg->data;
+
+	assert(msg->iov_count <= SM2_IOV_LIMIT);
+
+	memset(ctx->msg.desc, 0, msg->iov_count * sizeof(*msg->desc));
+	if (msg->desc) {
+		memcpy(ctx->msg.desc, &msg->desc,
+		       msg->iov_count * sizeof(*msg->desc));
+	}
+	memcpy(ctx->msg.msg_iov, msg->msg_iov,
+	       msg->iov_count * sizeof(*msg->msg_iov));
+	memcpy(ctx->msg.rma_iov, msg->rma_iov,
+	       msg->rma_iov_count * sizeof(*msg->rma_iov));
+
+	ctx->bytes_sent = 0;
+	ctx->bytes_acked = 0;
+	ctx->bytes_total = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+	ctx->msgs_in_flight = 0;
+	ctx->peer_gid = peer_gid;
+	ctx->op = op;
+	ctx->op_flags = op_flags;
+	ctx->status_flags = 0;
+}
+
+ssize_t sm2_rma_cmd_fill_sar_xfer(struct sm2_xfer_entry *xfer_entry,
+				  struct sm2_sar_ctx *ctx)
+{
+	ssize_t bytes_to_send, payload_used;
+	int ret;
+	void *src, *dest;
+	enum fi_hmem_iface iface;
+	uint64_t device;
+	struct sm2_rma_msg *msg;
+	struct sm2_cmd_sar_rma_msg *cmd_rma;
+	int jv;
+
+	msg = &ctx->msg;
+
+	xfer_entry->hdr.proto = sm2_proto_sar;
+	sm2_generic_format(xfer_entry, ctx->ep->gid, ctx->op, 0, msg->data,
+			   ctx->op_flags, msg->context);
+	cmd_rma = (struct sm2_cmd_sar_rma_msg *) xfer_entry->user_data;
+	memset(cmd_rma, 0, sizeof(*cmd_rma));
+	cmd_rma->sar_hdr.context = ctx;
+
+	payload_used = 0;
+	if (ctx->op == ofi_op_read_req) {
+		cmd_rma->sar_hdr.request_offset = ctx->bytes_requested;
+		ctx->bytes_requested += SM2_RMA_INJECT_SIZE;
+		if (ctx->bytes_requested >= ctx->bytes_total) {
+			ctx->bytes_requested = ctx->bytes_total;
+			cmd_rma->sar_hdr.proto_flags |=
+				FI_SM2_SAR_LAST_MESSAGE_FLAG;
+		}
+		memcpy(cmd_rma->rma_iov, msg->rma_iov,
+		       sizeof(cmd_rma->rma_iov));
+		ctx->msgs_in_flight++;
+
+		return FI_SUCCESS;
+	}
+
+	for (jv = 0; jv < SM2_IOV_LIMIT; jv++) {
+		if (ctx->bytes_sent == ctx->bytes_total)
+			break;
+		bytes_to_send =
+			MIN(msg->msg_iov[0].iov_len, msg->rma_iov[0].len);
+		bytes_to_send =
+			MIN(bytes_to_send, SM2_RMA_INJECT_SIZE - payload_used);
+		if (bytes_to_send == 0)
+			break;
+
+		src = (uint8_t *) msg->msg_iov[0].iov_base;
+		sm2_get_iface_device((void **) msg->desc, &iface, &device);
+		cmd_rma->rma_iov[jv].addr = msg->rma_iov[0].addr;
+		cmd_rma->rma_iov[jv].key = msg->rma_iov[0].key;
+		cmd_rma->rma_iov[jv].len = bytes_to_send;
+		dest = cmd_rma->user_data + payload_used;
+		assert(payload_used < SM2_INJECT_SIZE);
+
+		ret = ofi_copy_from_hmem(iface, device, dest, src,
+					 bytes_to_send);
+		if (ret)
+			return ret;
+		ofi_consume_iov_desc(msg->msg_iov, msg->desc, &msg->iov_count,
+				     bytes_to_send);
+		ofi_consume_rma_iov(msg->rma_iov, &msg->rma_iov_count,
+				    bytes_to_send);
+
+		payload_used += bytes_to_send;
+		ctx->bytes_sent += bytes_to_send;
+	}
+
+	if (ctx->bytes_sent == ctx->bytes_total) {
+		cmd_rma->sar_hdr.proto_flags |= FI_SM2_SAR_LAST_MESSAGE_FLAG;
+	}
+	ctx->msgs_in_flight++;
+
+	return FI_SUCCESS;
```

Comment:
for 0 byte operations, you are sending 1 message which I am guessing is for you to post a completion on the remote side. Do we want to do this?

Reply:
yes, we must send a message in order for remote to post completion.  For 0-byte reads I avoided it, but for writedata we couldn't.  Now with RMA_EVENT capability, we must always send the message for 0-sized operations.  In fact, I had to re-order the loop to `do {...} while()` so that we send even for zero-sized operations.